src/qibocal/protocols/readout_optimization/resonator_optimization.py

```diff
@@ -0,0 +1,330 @@
+    """Width for frequency sweep relative  to the readout frequency [Hz]."""
+    freq_step: int
+    """Frequency step for sweep [Hz]."""
+    amplitude_start: float
+    """Minimum amplitude."""
+    amplitude_stop: float
+    """Maximum amplitude."""
+    amplitude_step: float
+    """Step amplitude."""
+
+
+@dataclass
+class ResonatorOptimizationResults(Results):
+    """Resonator optimization outputs"""
+
+    fidelities: dict[QubitId, list]
+    """Assignment fidelities."""
+    best_freq: dict[QubitId, float]
+    """Resonator Frequency with the highest assignment fidelity."""
+    best_amp: dict[QubitId, list]
+    """Amplitude with lowest error."""
+    best_angle: dict[QubitId, float]
+    """IQ angle that maximes assignment fidelity."""
+    best_threshold: dict[QubitId, float]
+    """Threshold that maximes assignment fidelity."""
+
+
+ResonatorOptimizationType = np.dtype(
+    [
+        ("assignment_fidelity", np.float64),
+        ("avaraged_fidelity", np.float64),
+        ("frequency", np.float64),
+        ("amplitude", np.float64),
+        ("angle", np.float64),
+        ("threshold", np.float64),
+    ]
+)
+"""Custom dtype readout optimization."""
+
+
+@dataclass
+class ResonatorOptimizationData(Data):
+    """Data class for resonator optimization protocol."""
+
+    resonator_type: str
+    """Resonator type."""
+    amplitudes: dict[QubitId, float] = field(default_factory=dict)
+    """Amplitudes provided by the user."""
+    data: dict[QubitId, npt.NDArray[ResonatorOptimizationType]] = field(
+        default_factory=dict
+    )
+
+
+def _acquisition(
+    params: ResonatorOptimizationParameters,
+    platform: CalibrationPlatform,
+    targets: list[QubitId],
+) -> ResonatorOptimizationData:
+    r"""
+    Data acquisition for readout optimization.
+
+    Args:
+        params (ResonatorFrequencyParameters): experiment's parameters
+        platform (Platform): Qibolab platform object
+        qubits (list): list of target qubits to perform the action
+    """
+    delta_frequency_range = np.arange(
+        -params.freq_width / 2, params.freq_width / 2, params.freq_step
+    )
+
+    # taking advantage of multiplexing, apply the same set of gates to all qubits in parallel
+    ro_pulses_0 = {}
+    ro_pulses_1 = {}
+    amplitudes = {}
+    freq_sweepers = {}
+
+    sequence_0 = PulseSequence()
+    sequence_1 = PulseSequence()
+
+    for qubit in targets:
+        natives = platform.natives.single_qubit[qubit]
+
+        qd_channel, qd_pulse = natives.RX()[0]
+        ro_channel, ro_pulse_0 = natives.MZ()[0]
+        _, ro_pulse_1 = natives.MZ()[0]
+
+        ro_pulses_0[qubit] = ro_pulse_0
+        ro_pulses_1[qubit] = ro_pulse_1
+        amplitudes[qubit] = ro_pulse_0.probe.amplitude
+
+        sequence_0.append((ro_channel, ro_pulse_0))
+
+        sequence_1.append((qd_channel, qd_pulse))
+        sequence_1.append((ro_channel, Delay(duration=qd_pulse.duration)))
+        sequence_1.append((ro_channel, ro_pulse_1))
```

Comment:
```suggestion
        sequence_0 = natives.MZ()
        sequence_1 = natives.RX() | natives.MZ()
        amplitudes[qubit] = natives.MZ()[0][1].probe.amplitude
```
Similar to what I suggested for the `resonator_frequency`.